pkg/controllers/provisioning/launcher.go

```diff
@@ -38,24 +38,36 @@ import (
 )
 
 type Launcher struct {
-	Packer        *binpacking.Packer
-	KubeClient    client.Client
-	CoreV1Client  corev1.CoreV1Interface
-	CloudProvider cloudprovider.CloudProvider
+	Packer          *binpacking.Packer
+	KubeClient      client.Client
+	CoreV1Client    corev1.CoreV1Interface
+	CloudProvider   cloudprovider.CloudProvider
+	ResourceCounter *ResourceCounter
 }
 
-func (l *Launcher) Launch(ctx context.Context, schedules []*scheduling.Schedule, instanceTypes []cloudprovider.InstanceType) error {
+func (l *Launcher) Launch(ctx context.Context, schedules []*scheduling.Schedule,
+	instanceTypes []cloudprovider.InstanceType, provisioner v1alpha5.Provisioner) error {
+
 	// Pack and bind pods
 	errs := make([]error, len(schedules))
+	remainingResources, err := l.ResourceCounter.remainingResourceCounts(ctx, provisioner.Spec.Limits, provisioner.Name)
```

Comment:
I'm not sure how good of an idea this is, but WDYT about keeping track of launched resources in a separate controller that updates `provisioner.status.resources` and then we can check that without doing the math in the critical path.

We'd need to read the status of the provisioner on each loop, but it should hit the cache and be pretty quick. It opens us up to eventual consistency issues, but we already have them as you've described. 

It has a nice side effect of customers being able to see the data with `kubectl get provisioner`

Reply:
> I'm not sure how good of an idea this is, but WDYT about keeping track of launched resources in a separate controller that updates provisioner.status.resources and then we can check that without doing the math in the critical path.

That's a good idea. I tried caching things within the provisioning controller because I'm worried about listing all nodes in the cluster each provisioning loop (I've seen this call take 3-4 seconds for large clusters). I realized really quick that it breaks because the termination's happening in a different controller and the cache goes out of date but I should've broken out the control logic entirely.